src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs

```diff
@@ -69,11 +72,59 @@ protected override int GetFields<T>(T instance, LazyStringValue key, object docu
                 object value;
                 if (obj.IsObject() && obj.IsArray() == false)
                 {
+                    //Note that for dynamic fields 'field.Name' will contain the real name of the field and not 'property'
                     var val = TryDetectDynamicFieldCreation(property, obj.AsObject(), field);
                     if (val != null)
                     {
-                        value = GetValue(val);
-                        newFields += GetRegularFields(instance, field, value, indexContext);
+                        if (val.IsObject() && val.AsObject().TryGetValue("$spatial", out var dynamicSpatial))
+                        {
+                            obj = val;
+                        }
+                        else
+                        {
+                            value = GetValue(val);
+                            newFields += GetRegularFields(instance, field, value, indexContext);
+                            continue;
+                        }
+                    }
+
+                    if (obj.AsObject().TryGetValue("$spatial", out var inner))
+                    {
+                        if (inner.IsObject() == false)
+                        {
+                            throw new InvalidSpatialArgument($"Property {field.Name} is a spatial field but its value is not a valid spatial value.");
+                        }
+
+                        var sf = inner.AsObject();
+                        if (sf.TryGetValue("Type", out var type) == false || type.IsString() == false)
+                        {
+                            throw new InvalidSpatialArgument($"Property {field.Name} is a spatial field but it doesn't have a valid type.");
+                        }
+
+                        SpatialField spatialField = null;
+                        IEnumerable<AbstractField> spatial = null;
+                        switch (type.AsString())
```

Comment:
Do we need a type, we can just tell by the precence of fields, no?

Reply:
I guess we can avoid this extra field, but I'll still need to have the same number of parameters since there is no overloading in JS